### PR TITLE
disable plugin reload at every repair workflow start

### PR DESCRIPTION
**Pull request by kumarnareshh74**
**Issue**
FH loads all plugins during every repair workflow start. Plugin loading happens through reflection which is very memory intensive.

**Fix**
Load all plugins only once at service start and cache it. Use the cached plugins during repair workflow start. 

**Impact**
 - Customers not consuming plugins won't be impacted
 - Default behavior is not changed by this change. Customer must flip the newly introduced application parameter - **UsePluginModelV2** for the new behavior.

**Contract changes**
Below are the contract changes:
 - New application parameter - **UsePluginModelV2** has been introduced.
 - Attributes - **CustomServiceInitializer** and **RepairPredicateType** are unified to a single attribute - **Plugin**. Deprecation notice has been added to those 2 existing attributes.
 - All predicate types must implement a new interface IPredicateType.
 - Deprecation notice has been added to **RegisterToPredicateTypesCollection** API of **IRepairPredicateType** interface. Customers must start using **LoadPredicateTypes** - irrespective of the **UsePluginModelV2** application parameter value. This method must be implemented to avoid the compile time error.

### CODE REVIEW

File: FabricHealer/FabricHealer.cs [hunk 1]
```diff
@@ -42,6 +46,28 @@ protected override async Task RunAsync(CancellationToken cancellationToken)
             await healerManager.StartAsync();
         }
 
+        private async Task LoadPluginsAsync(CancellationToken cancellationToken)
+        {
+            if (!FabricHealerManager.ConfigSettings.EnableCustomRepairPredicateType &&
```
Comment (sidhant012) [inline FabricHealer/FabricHealer.cs:51]: why is this check needed if the LoadPluginsAsync() method is only called when the UsePluginModelV2 config is true

File: FabricHealer/PackageRoot/Config/LogicRules/AppRules.guan [hunk 2]
```diff
@@ -201,4 +201,4 @@ TimeScopedRestartReplica :- RestartReplica(DoHealthChecks=false, MaxWaitTimeForH
 
 
 ## This is an example of how to call a custom repair from a plugin. In this rule, "SampleRepair" is the repair rule loaded from SampleHealerPlugin project.
-##Mitigate(HealthState="Warning") :- SampleRepair("repair rule loaded from plugin").
```
Comment (sidhant012) [inline FabricHealer/PackageRoot/Config/LogicRules/AppRules.guan:204]: There is no change here? Why is this file committed?

Comment (kumarnareshh74) [reply FabricHealer/PackageRoot/Config/LogicRules/AppRules.guan]: whitespace. 
reverted it now

File: FabricHealer/FabricHealer.cs (hunk 1 above)
Comment (kumarnareshh74) [reply FabricHealer/FabricHealer.cs:51]: checking against bad configuration. 

if there are no custom service initializer plugins and repair predicate plugins, we dont wanna load any plugins.

**Comment by sidhant012** [comment]
We shouldn't have user config settings for something like this. Model should work similar to how FabricObserver does. If there are plugins in the folder, then there are plugins... There is no custom settings for FH plugins beyond enabling and then choosing an interface.

**Comment by sidhant012** [comment]
I dont think we should be catching null reference exceptions in BasePluginLoader

**Comment by GitTorre** [comment]
The initial implemenation was wrong in that plugins were loaded each time the Guan predicate loader ran. I do not understand why we would want to keep that impl around and hide with a user configuration for a new interface (V2). Please remove the old implementation and the new settings related to its usage.

**Comment by kumarnareshh74** [comment]
> BasePluginLoader

this is old code. As it'll be removed as a whole, didnt want to make the PR more complex by doing those changes

**Comment by kumarnareshh74** [comment]
> We shouldn't have user config settings for something like this. Model should work similar to how FabricObserver does. If there are plugins in the folder, then there are plugins... There is no custom settings for FH plugins beyond enabling and then choosing an interface.

this is just like a feature flag. In worst case, it would be helpful. 
Not needed. Added just for safety

**Comment by kumarnareshh74** [comment]
> The initial implementation was wrong in that plugins were loaded each time the Guan predicate loader ran. I do not understand why we would want to keep that impl around and hide with a user configuration for a new interface (V2). Please remove the old implementation and the new settings related to its usage.

1 - old impl. is functionally correct. But it performs very poor.
2 - kept the old impl. just to avoid breaking changes. But if you are fine, then I'll do that

**Comment by kumarnareshh74** [comment]
> > The initial implementation was wrong in that plugins were loaded each time the Guan predicate loader ran. I do not understand why we would want to keep that impl around and hide with a user configuration for a new interface (V2). Please remove the old implementation and the new settings related to its usage.
> 
> 1 - old impl. is functionally correct. But it performs very poor. 2 - kept the old impl. just to avoid breaking changes. But if you are fine, then I'll do that

removed the old code

**Comment by kumarnareshh74** [comment]
> > BasePluginLoader
> 
> this is old code. As it'll be removed as a whole, didnt want to make the PR more complex by doing those changes

removed the old code

File: FabricHealer/Utilities/PluginLoader.cs [hunk 3]
```diff
@@ -3,38 +3,70 @@
...
+        private static readonly IDictionary<IRepairPredicateType, IList<PredicateType>> PluginToPredicateTypesMap = new Dictionary<IRepairPredicateType, IList<PredicateType>>();
+        private static readonly HashSet<ICustomServiceInitializer> CustomServiceInitializers = new();
 
+        public FabricHealerPluginLoader(ServiceContext context)
+        {
+            this._serviceContext = context;
+        }
 
-        protected BasePluginLoader(Logger logger, ServiceContext serviceContext)
+        public async Task InitializePluginsAsync(CancellationToken cancellationToken)
```
Comment (GitTorre) [inline FabricHealer/Utilities/PluginLoader.cs:27]: This should be static.

File: FabricHealer/Utilities/PluginLoader.cs [hunk 4]
```diff
@@ -3,38 +3,70 @@
...
-            this.ServiceContext = serviceContext;
+            foreach (var customServiceInitializer in FabricHealerPluginLoader.CustomServiceInitializers)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                await customServiceInitializer.InitializeAsync();
+            }
         }
 
-        protected virtual Object GetPluginClassInstance(Assembly assembly)
+        public void LoadPluginPredicateTypes()
```
Comment (GitTorre) [inline FabricHealer/Utilities/PluginLoader.cs:36]: This should be static.

File: FabricHealer/Utilities/PluginLoader.cs [hunk 5]
```diff
@@ -3,38 +3,70 @@
...
+                foreach (var predicateType in predicateTypes)
+                {
+                    FabricHealerPluginLoader.PluginToPredicateTypesMap[repairPredicateType].Add((PredicateType)predicateType);
+                }
+            }
         }
 
-        protected virtual Task CallCustomAction(Object instance)
+
+        public void RegisterPredicateTypes(FunctorTable functorTable, string serializedRepairData)
```
Comment (GitTorre) [inline FabricHealer/Utilities/PluginLoader.cs:51]: This should be static.

File: FabricHealer/FabricHealer.cs [hunk 6]
```diff
@@ -33,19 +33,32 @@ protected override async Task RunAsync(CancellationToken cancellationToken)
...
         {
-            var pluginLoader = new ServiceInitializerPluginLoader(this.logger, this.Context);
-            await pluginLoader.LoadPluginsAndCallCustomAction(typeof(CustomServiceInitializerAttribute), typeof(ICustomServiceInitializer));
+            if (!FabricHealerManager.ConfigSettings.EnableCustomRepairPredicateType &&
+                !FabricHealerManager.ConfigSettings.EnableCustomServiceInitializers)
+            {
+                return;
+            }
+
+            var pluginLoader = new FabricHealerPluginLoader(this.Context);
```
Comment (GitTorre) [inline FabricHealer/FabricHealer.cs:50]: LoadPlugins is the only function that requires a service context. Why not just pass it into a static LoadPlugins? There is really no need for instance type here.

**Comment by GitTorre** [comment]
I made some changes.

File: FabricHealer/FabricHealer.cs (hunk 6 above)
Comment (kumarnareshh74) [reply FabricHealer/FabricHealer.cs]: i see this is done as part of your commit

File: FabricHealer/Utilities/PluginLoader.cs (hunk 5 above)
Comment (kumarnareshh74) [reply FabricHealer/Utilities/PluginLoader.cs]: i see this is done as part of your commit

File: FabricHealer/Utilities/PluginLoader.cs (hunk 4 above)
Comment (kumarnareshh74) [reply FabricHealer/Utilities/PluginLoader.cs]: i see this is done as part of your commit

File: FabricHealer/Utilities/PluginLoader.cs (hunk 3 above)
Comment (kumarnareshh74) [reply FabricHealer/Utilities/PluginLoader.cs]: i see this is done as part of your commit

File: FabricHealer/FabricHealer.cs (hunk 1 above)
Comment (sidhant012) [reply FabricHealer/FabricHealer.cs:41]: I see. Can you please update the plugin documentation files with this information like this and any other relevant changes so that the documentation for the two types of plugins are up to date.

File: FabricHealer.sln [hunk 7]
```diff
@@ -97,8 +97,8 @@ Global
 		{FB6B2E0E-2243-4447-BA56-7D91CB0D2127}.Release|x64.Build.0 = Release|x64
 		{88C2DBDC-2C91-4145-804C-B91879BDCB94}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{88C2DBDC-2C91-4145-804C-B91879BDCB94}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{88C2DBDC-2C91-4145-804C-B91879BDCB94}.Debug|x64.ActiveCfg = Debug|Any CPU
-		{88C2DBDC-2C91-4145-804C-B91879BDCB94}.Debug|x64.Build.0 = Debug|Any CPU
+		{88C2DBDC-2C91-4145-804C-B91879BDCB94}.Debug|x64.ActiveCfg = Debug|x64
+		{88C2DBDC-2C91-4145-804C-B91879BDCB94}.Debug|x64.Build.0 = Debug|x64
```
Comment (sidhant012) [inline FabricHealer.sln:101]: Why was this changed from Any CPU to x64?

File: FabricHealer/FabricHealer.cs (hunk 1 above)
Comment (sidhant012) [reply FabricHealer/FabricHealer.cs:41]: Still needs to be done. Will sync to discuss
